### ISSUE TITLE
omelasticsearch silent failure

### ISSUE THREAD

**Issue by n2N8Z** [rsyslog/rsyslog#2757]
### Expected behavior
Logs are either inserted, or an error is reported.

### Actual behavior
A whole dequeued batch can silently fail.

### Steps to reproduce the behavior
Unfortunately I don't have steps to reproduce.
I ran a test import of multiple files which together equal around 10M events. After the import finished the count in ES was exactly 5000 less than the count of the events in the files. 5000 is also the dequeue batch size. I extracted the ES events out, and compared them back to the log files, and the 5000 missing events are a contiguous block in the log files, supporting the idea that a whole dequeued batch went missing. During this run, no error logs were written to syslog, and the configured errorFile was not written. So the batch failed entirely silently.

Is there any other way that rsyslog could have lost a batch, other than due to an ES error that rsyslog did not report ?

When I repeated the test, the count of events in ES exactly matched the number in the logs files.

### Environment
- rsyslog version: 
rsyslogd 8.33.1, compiled with:
        PLATFORM:                               x86_64-redhat-linux-gnu
        PLATFORM (lsb_release -d):
        FEATURE_REGEXP:                         Yes
        GSSAPI Kerberos 5 support:              Yes
        FEATURE_DEBUG (debug build, slow code): No
        32bit Atomic operations supported:      Yes
        64bit Atomic operations supported:      Yes
        memory allocator:                       system default
        Runtime Instrumentation (slow code):    No
        uuid support:                           Yes
        systemd support:                        Yes
        Number of Bits in RainerScript integers: 64

See http://www.rsyslog.com for more information.

- platform: Red Hat Enterprise Linux Server release 7.5 (Maipo)
- for configuration questions/issues, include rsyslog.conf and included config files
The config is basically this:
global(workDirectory="/var/lib/rsyslog/perf_test" maxMessageSize="32k")
module(load="imfile" pollingInterval="10")

ruleset(name="rule1" queue.type="LinkedList" queue.filename="rule1" queue.saveonshutdown="on" queue.workerthreads="4" queue.dequeuebatchsize="5000")
{
  action(type="omelasticsearch" server="es" dynSearchIndex="on" searchIndex="perf_test" template="temp1" action.resumeretrycount="-1" errorFile="/var/log/rsyslog-es-error.perftest.log" bulkmode="on" bulkId="UUID" dynBulkId="on")
}

input(type="imfile"
  ruleset="rule1"
  file="/perftest/*"
  persistStateInterval="10"
  tag=""
  severity="debug"
  startmsg.regex="^regex..."
  readTimeout="3"
)

**Comment by davidelang** [comment]
did the messages end up in the errorfile?

**Comment by rgerhards** [comment]
I just want to set the expectations straight: as this seems not reproducible, chances are around 0 that we can fix it. I'll keep it open in case someone actually wants to try to tackle this.

**Comment by n2N8Z** [comment]
Is there any other way that rsyslog could have lost a batch, other than due to an ES error that rsyslog did not report ?

I did have quite a lot of lost events previously when not using the bulk api, with an earlier version of rsyslog, and with around 20 worker threads configured. Perhaps I can go back to that and try to find some more details, or perhaps try to get a debug capture of it happening.

**Comment by rgerhards** [comment]
> Is there any other way that rsyslog could have lost a batch, other than due to an ES error that rsyslog did not report ?

not that I am aware of

**Comment by richm** [comment]
enable impstats - there are many omelasticsearch impstats, some of which are related to errors

**Comment by n2N8Z** [comment]
@richm: yes, I had impstats enabled. So you are suggesting that there would be errors recorded in impstats, which are produced by rsyslog, and yet rsyslog does not consider them to be errors worthy of writing to the configured errorFile ?
As I had stated: "no error logs were written to syslog, and the configured errorFile was not written. So the batch failed entirely silently."

**Comment by richm** [comment]
> So you are suggesting that there would be errors recorded in impstats, which are produced by rsyslog, and yet rsyslog does not consider them to be errors worthy of writing to the configured errorFile ?

It's worth checking, especially if you already have impstats enabled: https://www.rsyslog.com/doc/v8-stable/configuration/modules/omelasticsearch.html#omelasticsearch-statistic-counter

**Comment by rgerhards** [comment]
Sorry for the very long delay. rsyslog does not lightly close issues and I am not closing this just because it is old; I checked the retained evidence and current test/doc coverage while working through the backlog.

This report was serious, but it was explicitly non-reproducible even at the time: the same import succeeded on retry, and no reproducer or debug capture was later added. Current omelasticsearch has substantially more bulk failure handling coverage than the 8.33.1-era report retained here, including tests for bulk error-file population and bulk retry behavior, and the current documentation describes the failure counters and retry handling. Without a reproducer, exact Elasticsearch response, debug log, or current-version confirmation, there is no actionable path left in this issue.

I am closing this as not reproducible with the available evidence. If anyone sees silent omelasticsearch batch loss on a current rsyslog version, please open a new issue with a minimal config, Elasticsearch/mock-server response details, impstats counters, errorFile setting/result, and a reproducer if possible. Data-loss reports are worth reopening with current evidence.
